app/src/ui/lib/tooltip.tsx

```diff
@@ -0,0 +1,533 @@
+      show: false,
+      targetRect: new DOMRect(),
+      hostRect: new DOMRect(),
+      windowRect: new DOMRect(),
+      tooltipRect: new DOMRect(),
+      tooltipHost: tooltipHostFor(target),
+    }
+
+    this.resizeObserver = new ResizeObserver(entries => {
+      for (const entry of entries) {
+        if (entry.target === this.tooltipRef) {
+          const tooltipRect = this.tooltipRef.getBoundingClientRect()
+          if (!rectEquals(this.state.tooltipRect, tooltipRect)) {
+            this.setState({ tooltipRect })
+          }
+        } else if (entry.target === this.state.target) {
+          const targetRect = this.state.target.getBoundingClientRect()
+          if (!rectEquals(this.state.targetRect, targetRect)) {
+            this.setState({ targetRect })
+          }
+        }
+      }
+    })
+  }
+
+  public componentDidMount() {
+    const { target } = this.props
+    target.subscribe(this.onTargetRef)
+
+    if (target.current !== null && this.state.target !== target.current) {
+      this.onTargetRef(target.current)
+    } else if (this.state.target !== null) {
+      this.installTooltip(this.state.target)
+    }
+  }
+
+  public onTargetRef = (target: TooltipTarget | null) => {
+    this.setState({ target, tooltipHost: tooltipHostFor(target) })
+  }
+
+  public onTooltipRef = (elem: HTMLDivElement | null) => {
+    if (elem === null) {
+      if (this.state.id) {
+        releaseUniqueId(this.state.id)
+        this.setState({ id: undefined })
+      }
+      this.resizeObserver.disconnect()
+      const { tooltipRef } = this
+
+      if (tooltipRef !== null) {
+        tooltipRef.removeEventListener('mouseenter', this.onTooltipMouseEnter)
+        tooltipRef.removeEventListener('mouseleave', this.onTooltipMouseLeave)
+      }
+    } else {
+      // The tooltip has been mounted, let's measure it and show it
+      this.setState({
+        tooltipRect: elem.getBoundingClientRect() ?? new DOMRect(),
+        show: true,
+        measure: false,
+        id: this.state.id ?? createUniqueId('tooltip'),
+      })
+      this.resizeObserver.observe(elem)
+      if (this.props.interactive === true) {
+        elem.addEventListener('mouseenter', this.onTooltipMouseEnter)
+        elem.addEventListener('mouseleave', this.onTooltipMouseLeave)
+      }
+    }
+
+    this.tooltipRef = elem
+  }
+
+  private onTooltipMouseEnter = (e: MouseEvent) => this.cancelHideTooltip()
+  private onTooltipMouseLeave = (e: MouseEvent) => this.beginHideTooltip()
+
+  public componentDidUpdate(
+    prevProps: ITooltipProps<T>,
+    prevState: ITooltipState
+  ) {
+    const { target } = this.state
+
+    if (prevProps.target !== this.props.target) {
+      prevProps.target.unsubscribe(this.onTargetRef)
+      this.props.target.subscribe(this.onTargetRef)
+    }
+
+    if (target !== prevState.target) {
+      if (target === null) {
+        this.removeTooltip(prevState.target)
+      } else {
+        this.installTooltip(target)
+      }
```

Comment:
Shouldn't we always remove the tooltip from the previous target?
```suggestion
      this.removeTooltip(prevState.target)
      if (target !== null) {
        this.installTooltip(target)
      }
```

Reply:
Yeah, good catch!